File: src/lib/search-operators.ts
/**
 * SearchPanel query operators. Parses `tag:`, `path:`, and the
 * free-text remainder out of a search query. The plain-text part is
 * forwarded to the existing Rust `searchVault` command; the operators
 * are evaluated client-side against the tag/link/vault indices.
 *
 * Syntax:
 *   `tag:#foo`            — files carrying the #foo tag (literal `#` optional)
 *   `tag:projects/markup` — nested tag
 *   `path:journal/`       — paths whose vault-relative form contains the substring
 *   anything else         — added to the free-text query
 *
 * Operators are case-insensitive on their values for `path:`; tag
 * values are exact (case preserved per Obsidian convention).
 */

export interface ParsedQuery {
  /** Tag values (without leading `#`). Empty when no tag operator. */
  tags: string[];
  /** Path substrings. Empty when no path operator. */
  paths: string[];
  /** Free-text remainder; everything outside operators. May be empty. */
  text: string;
}

const OPERATOR_RE = /\b(tag|path):(\S+)/gi;

export function parseQuery(raw: string): ParsedQuery {
  if (!raw) return { tags: [], paths: [], text: "" };
  const tags: string[] = [];
  const paths: string[] = [];
  let text = raw;
  text = text.replace(OPERATOR_RE, (_m, kind, value) => {
    const k = kind.toLowerCase();
    if (k === "tag") {
      tags.push(value.replace(/^#/, ""));
    } else if (k === "path") {
      paths.push(value);
    }
    return ""; // strip the operator token from the free-text part
  });
  return { tags, paths, text: text.replace(/\s+/g, " ").trim() };
}

/** True when `path` matches every path-operator substring (case-
 *  insensitive). Returns true when there are no path operators. */
export function pathMatches(path: string, pathOps: string[]): boolean {
  if (pathOps.length === 0) return true;
  const lower = path.toLowerCase();
  return pathOps.every((p) => lower.includes(p.toLowerCase()));
}

File: ios/MarkupKit/Sources/MarkupKit/SearchOperators.swift
import Foundation

/// The result of splitting a search query into operators + free text.
/// Mirrors the desktop `ParsedQuery` from `src/lib/search-operators.ts`.
public struct ParsedQuery: Equatable, Sendable {
    /// Tag values (without a leading `#`). Empty when no `tag:` operator.
    public var tags: [String]
    /// Path substrings. Empty when no `path:` operator.
    public var paths: [String]
    /// Free-text remainder; everything outside operators. May be empty.
    public var text: String

    public init(tags: [String], paths: [String], text: String) {
        self.tags = tags
        self.paths = paths
        self.text = text
    }
}

private let operatorRegex = try! NSRegularExpression(
    pattern: "\\b(tag|path):(\\S+)", options: [.caseInsensitive])

/// Parse `tag:` / `path:` operators and the free-text remainder out of a query.
///
///   `tag:#foo`            — files carrying the #foo tag (literal `#` optional)
///   `path:journal/`       — paths whose vault-relative form contains the substring
///   anything else         — added to the free-text query
///
/// Faithful port of `parseQuery` in `src/lib/search-operators.ts`.
public func parseQuery(_ raw: String) -> ParsedQuery {
    if raw.isEmpty { return ParsedQuery(tags: [], paths: [], text: "") }

    var tags: [String] = []
    var paths: [String] = []
    let ns = raw as NSString
    let fullRange = NSRange(location: 0, length: ns.length)

    for m in operatorRegex.matches(in: raw, range: fullRange) {
        let kind = ns.substring(with: m.range(at: 1)).lowercased()
        let value = ns.substring(with: m.range(at: 2))
        if kind == "tag" {
            tags.append(value.hasPrefix("#") ? String(value.dropFirst()) : value)
        } else if kind == "path" {
            paths.append(value)
        }
    }

    var text = operatorRegex.stringByReplacingMatches(
        in: raw, range: fullRange, withTemplate: "")
    text = text.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
    text = text.trimmingCharacters(in: .whitespacesAndNewlines)

    return ParsedQuery(tags: tags, paths: paths, text: text)
}

/// True when `path` matches every `path:` operator substring (case-insensitive).
/// Returns true when there are no path operators.
public func pathMatches(_ path: String, _ pathOps: [String]) -> Bool {
    if pathOps.isEmpty { return true }
    let lower = path.lowercased()
    return pathOps.allSatisfy { lower.contains($0.lowercased()) }
}
